targets/TARGET_NXP/TARGET_LPC176X/device/CRP.c

```diff
@@ -0,0 +1,60 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	 http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 
+ #include "mbed_toolchain.h"
+
+ /* Code Read Protection
+
+NONE	0xFFFFFFFF - No code protection.
+
+CRP1    0x12345678 - Write to RAM command can not access RAM below 0x10000200.
+                   - Read Memory command: disabled.
+                   - Copy RAM to Flash command: cannot write to Sector 0.
+                   - "Go" command: disabled.
+                   - Erase sector(s) command: can erase any individual sector except 
+                   	 sector 0 only, or can erase all sectors at once.
+                   - Compare command: disabled
+
+CRP2    0x87654321 - Write to RAM command: disabled.
+                   - Copy RAM to Flash: disabled.
+                   - Erase command: only allows erase of all sectors.
+
+CRP3    0x43218765 - Access to chip via the SWD pins is disabled. ISP entry
+                     by pulling PIO0_1 LOW is disabled if a valid user code is
+                     present in flash sector 0.
+Caution: If CRP3 is selected, no future factory testing can be
+performed on the device.
+*/
+#if !defined(APPLICATION_ADDR) // Relocate CRP if there is a bootloader.
+    #define APPLICATION_ADDR 0
+#endif
+
+#define CRP_NONE	0xFFFFFFFF
+#define CRP_1		0x12345678 
+#define CRP_2		0x87654321
+#define CRP_3		0x43218765
+
+#ifndef CRP
+#define CRP CRP_NONE
+#endif
+
+#if defined (__ICCARM__)
+    __root const long CRP_Key@APPLICATION_ADDR + 0x000002FC = CRP;
+#elif defined (__ARMCC_VERSION)
+    const long CRP_Key __attribute__((used)) __attribute__((at(APPLICATION_ADDR+0x000002FC))) = CRP;
+#elif defined (__GNUC__)
+    const long CRP_Key __attribute__((used)) __attribute__((section(".CRPSection"))) = CRP;
```

Comment:
You could simplify this by handling placement in the linker scripts. This could then be simplified to something like this for all the toolchains:
```c
MBED_SECTION(".CRPSection") const long CRP_Key = CRP;
```

The scatter file changes would look something like this:
```
ER_IROM0 MBED_APP_START 0x2FC  {  ; load address = execution address
    *.o (RESET, +First)
    .ANY (+RO)
}
ER_CRP (MBED_APP_START + 0x2FC) 4)  {
    *.o (.CRPSection)
}
ER_IROM1 (MBED_APP_START + (0x2FC + 4)) (MBED_APP_SIZE - (0x2FC + 4))  {
    *(InRoot$$Sections)
    .ANY (+RO)
}
```

Reply:
@c1728p9  I tried your suggestion and get the following
```
[ERROR] Error: L6788E: Scatter-loading of execution region ER_IROM1 to execution address [0x00010300,0x0001737c) will cause the contents of execution region ER_IROM1 at load-address [0x000100c4,0x00017140) to be corrupted at run-time.
Error: L6202E: __main.o(!!!main) cannot be assigned to non-root region 'ER_IROM1'
Error: L6202E: __scatter.o(!!!scatter) cannot be assigned to non-root region 'ER_IROM1'
Error: L6202E: __dczerorl.o(!!dczerorl) cannot be assigned to non-root region 'ER_IROM1'
Error: L6202E: __scatter_copy.o(!!handler_copy) cannot be assigned to non-root region 'ER_IROM1'
Error: L6202E: __scatter_zi.o(!!handler_zi) cannot be assigned to non-root region 'ER_IROM1'
Error: L6202E: anon$$obj.o(Region$$Table) cannot be assigned to non-root region 'ER_IROM1'
Error: L6203E: Entry point (0x00010301) lies within non-root region ER_IROM1.
```

Any assistance appreciated, I'm no expert on linker files!

Reply:
From what I can tell 'InRoot$$Sections' can only be placed in a root section, as it needs to place some ARM library stuff there. The root section here would be ER_IROM0, instead of ER_IROM1.
Moving it over might solve the issue?

Then again, I have no expert knowledge on this either.

Reply:
That gets rid of most errors but leaves me with 
```
ER_IROM0 MBED_APP_START 0x2FC  {  ; load address = execution address
    *.o (RESET, +First)
    *(InRoot$$Sections)
    .ANY (+RO)
}
ER_CRP (MBED_APP_START + 0x2FC) 4  {
    *.o (.CRPSection)
}
ER_IROM1 (MBED_APP_START + (0x2FC + 4)) (MBED_APP_SIZE - (0x2FC + 4))  {
    .ANY (+RO)
}
```
Error: L6788E: Scatter-loading of execution region ER_IROM1 to execution address 
 0x00010300,0x00017d80) will cause the contents of execution region ER_IROM1 at load-address [0x000101a4,0x00017c24) to be corrupted at run-time.
```

Reply:
It appears to be overwriting itself, though I can't really see how...

> This occurs when scatter-loading takes place and an execution region is put in a position where it partially or wholly overwrites another execution region (which can be itself or another region).
> For example, the following code generates this error:
> ```
> LOAD_ROM 0x0000 0x4000 {
>     EXEC1 0x4000 0x4000 {
>         *(+RW,+ZI)
>     }
>     EXEC2 0x0000 0x4000 {
>         *(+RO)
>     }
> }
> ```
> This code does not generate the error:
> ```
> LOAD_ROM 0x0000 0x4000
> {
>  EXEC1 0x0000 0x4000
>  {
>  * (+RO)
>  }
>  EXEC2 0x4000 0x4000
>  {
>  * (+RW,+ZI)
>  }
> }

```
Source: https://developer.arm.com/docs/dui0807/f/linker-errors-and-warnings/list-of-the-armlink-error-and-warning-messages

Reply:
Try adding the FIXED keyword to ER_CRP. This will force it to be at the right address
```
ER_CRP (MBED_APP_START + 0x2FC) FIXED 4  {
    *.o (.CRPSection)
}
```

Reply:
It doesn't seem to be related to the CRP section as removing it still causes
```
Error: L6788E: Scatter-loading of execution region ER_IROM1 to execution address [0x00010300,0x00017d80) will cause the contents of execution region ER_IROM1 at load-address [0x000101a4,0x00017c24) to be corrupted at run-time.
```

Reply:
You can try and use AssignExpr() to make sure ER_IROM 1's execution address cannot overlap with its load address, though I don't know if it will make a difference.
```
ER_IROM0 MBED_APP_START 0x2FC  {  ; load address = execution address
    *.o (RESET, +First)
    *(InRoot$$Sections)
    .ANY (+RO)
}
ER_CRP (MBED_APP_START + 0x2FC) 4  {
    *.o (.CRPSection)
}
ER_IROM1 AlignExpr(+0, 0x2000)( (MBED_APP_SIZE - (0x2FC + 4))  { ; Align it at the end of the previous section but set its execution address to 0x20000
    .ANY (+RO)
}
```
The code above obviously needs a better address then the hardcoded 0x20000 but it might work in your situation.

Reply:
Try this:
```
ER_IROM0 MBED_APP_START 0x2FC  {  ; load address = execution address
    *.o (RESET, +First)
    .ANY (+RO)
}
ER_CRP (MBED_APP_START + 0x2FC) FIXED 4  {
    *.o (.CRPSection)
}
ER_IROM1 (MBED_APP_START + (0x2FC + 4)) FIXED (MBED_APP_SIZE - (0x2FC + 4))  {
    *(InRoot$$Sections)
    .ANY (+RO)
}
```